### PR TITLE
mgr/cephadm: Cleanup snmp

### PR DESCRIPTION
**Pull request by sebastian-philipp**
## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/python-common/ceph/deployment/service_spec.py [hunk 1]
```diff
@@ -1199,24 +1196,34 @@ def __init__(self,
...
         self.privacy_protocol = privacy_protocol
 
+    @classmethod
+    def _from_json_impl(cls, json_spec: dict) -> 'SNMPGatewaySpec':
+
+        cpy = json_spec.copy()
+        types = [
+            ('snmp_version', SNMPGatewaySpec.SNMPVersion),
+            ('auth_protocol', SNMPGatewaySpec.SNMPAuthType),
+            ('privacy_protocol', SNMPGatewaySpec.SNMPPrivacyType),
```
Comment (sebastian-philipp) [inline src/python-common/ceph/deployment/service_spec.py:1206]: @adk3798 we should move this table over to the base class anc combine it with 

https://github.com/ceph/ceph/pull/44148/commits/9742e950c8ded9169ad3d07aa396e8607b8aa44b#diff-91ff13f38df816b588aedff4193f02803771e1702b8dabd8f67b3e36d939e570R767-R774

IMO there is no need that we implement this  `_from_json_impl` for every subclass.

Comment (adk3798) [reply src/python-common/ceph/deployment/service_spec.py:1206]: yeah, we don't have `_from_json_impl` for any of the other services spec classes and there's nothing bout this service I see that should make use have to change that. +1 to combing it with KNOWN_PROPERTIES

File: src/python-common/ceph/deployment/service_spec.py [hunk 2]
```diff
@@ -1216,7 +1216,7 @@ def _from_json_impl(cls, json_spec: dict) -> 'SNMPGatewaySpec':
 
     @property
     def ports(self) -> List[int]:
-        return [self.port]
+        return [self.port or 9464]
```
Comment (pcuzner) [inline src/python-common/ceph/deployment/service_spec.py:1219]: Combined with the other changes, does it make to also avoid the explicit definition here and pick it up from the class?

File: src/python-common/ceph/deployment/service_spec.py [hunk 3]
```diff
@@ -1227,16 +1234,6 @@ def _check_type(name: str, value: Optional[str], options: List[str]) -> None:
                 'Missing SNMP version (snmp_version)'
             )
 
-        _check_type('snmp_version',
-                    self.snmp_version,
-                    list(set(opt.value for opt in SNMPVersion)))
-        _check_type('auth_protocol',
```
Comment (pcuzner) [inline src/python-common/ceph/deployment/service_spec.py:1233]: I can see the tests passing, but when I actually try and deploy bogus auth_protocol or privacy settings don't get caught. Maybe it's my local environment? Have you tested a deployment with bogus settings i.e. ceph orch apply -i <file> 
<pre>
service_type: snmp-gateway
service_name: snmp-gateway
placement:
  count: 1
spec:
  credentials:
    snmp_v3_auth_password: mypassword
    snmp_v3_auth_username: myuser
    snmp_v3_priv_password: mysecret
  port: 9464
  engine_id: 8000c53f0000000000
  privacy_protocol: WEIRD
  snmp_destination: 192.168.122.1:162
  auth_protocol: BIZARRE
  snmp_version: V3
</pre>

File: src/python-common/ceph/deployment/service_spec.py (hunk 2 above)
Comment (sebastian-philipp) [reply src/python-common/ceph/deployment/service_spec.py:1229]: right now, this is similar to how we do things in MonitoringSpec

**Comment by pcuzner** [comment]
My local tests are passing now - thanks! Once the tox issue is resolved, happy to approve.

**Comment by sebastian-philipp** [comment]
https://pulpito.ceph.com/swagner-2022-01-03_15:37:23-orch:cephadm-wip-swagner-testing-2022-01-03-1230-distro-default-smithi/

* Failures are due to the environment

**Comment by sebastian-philipp** [comment]
```
cephadm/tests/test_spec.py:11:1: F401 'ceph.deployment.hostspec.SpecValidationError' imported but unused
1     F401 'ceph.deployment.hostspec.SpecValidationError' imported but unused
```

**Comment by sebastian-philipp** [comment]
changelog: removed the single import line in `cephadm/tests/test_spec.py`

**Review by adk3798 — APPROVED** [review]

**Pull request merged**